Sort keys when encoding tables as JSON

winniehell opened the pull request:
## Description

Sorts object keys when encoding a table as JSON. This makes JSON objects deterministic which is especially useful for the GLightbox parameters which are part of the rendered HTML. This change may have unintended side effects such as performance degradation which I hope will be revealed by a review.

closes https://github.com/quarto-dev/quarto-cli/issues/11122

## Checklist

I have (if applicable):

- [x] filed a [contributor agreement](https://github.com/quarto-dev/quarto-cli/blob/main/CONTRIBUTING.md).
- [x] referenced the GitHub issue this PR closes
- [x] updated the appropriate changelog

winniehell committed:
Sort keys when encoding tables as JSON

winniehell commented:
I verified locally that this change resolves https://github.com/quarto-dev/quarto-cli/issues/11122 but of course it has larger impact on the code base which I'm not sure is wanted.

I will still send in the contributor agreement later.

cscheid commented:
Thanks for the PR. The fix makes sense, but `json.lua` is an upstream dependency, see the [header](https://github.com/quarto-dev/quarto-cli/pull/11124/files#diff-4119ff37fe985fddcc6bf731212d9406032c9177300ea25ecd38339bf30de825R5):

```lua
--
-- json.lua
--
-- Copyright (c) 2020 rxi
-- https://github.com/rxi/json.lua
--
```

I have a slight preference towards making a fix upstream first.

winniehell commented:
@cscheid I can try but that library has not seen any activity in the last four years, so I'm not optimistic it will get merged any time soon.

cscheid commented:
I'd be more willing to take a PR here if the fix were mergeable upstream first. Your fix creates a dependency between json.lua and Quarto, and that's not really wanted in this context.

winniehell commented:
yep, I understand. :+1: the alternative however implies that `sortedPairs` is duplicated upstream which leads to at least two copies in Quarto. There are already multiple anyway so this may not be a big deal.

cscheid commented:
We're not a go shop but the [appropriate proverb](https://go-proverbs.github.io/) applies here: [A little copying is better than a little dependency.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s)

winniehell commented:
We are in agreement here and for this situation and I also dislike big dependency trees. But I have also made bad experience with legacy code bases that had a lot of duplication and some of the duplicates were adjusted to behave slightly different. So there is probably a sweet spot somewhere in the middle (and just to be absolutely clear: I'm not suggesting any change in strategy for Quarto here, just sharing my point of view).

winniehell commented:
upstream pull request: https://github.com/rxi/json.lua/pull/51

winniehell committed:
Update json.lua from current (unreleased) master

 https://github.com/rxi/json.lua/blob/dbf4b2dd2eb7c23be2773c89eb059dadd6436f94/json.lua

winniehell committed:
Update json.lua from pending pull request

https://github.com/rxi/json.lua/pull/51

winniehell committed:
Update json.lua from pending pull request

https://github.com/rxi/json.lua/pull/52

winniehell commented:
@cscheid yep, I have also included the other unreleased changes from upstream and created a pull request for your earlier changes: https://github.com/rxi/json.lua/pull/52

winniehell commented on src/resources/pandoc/datadir/_json.lua:
```diff
@@ -101,13 +116,13 @@ local function encode_table(val, stack)
     return "{" .. table.concat(res, ",") .. "}"
   else
-    return "[]"
+    error( string.format("invalid table: unsupported key type %s", types[1]) )
```
change in functionality to maintain compatibility with upstream tests, encoding `{ [ function() end ] = 12 }` in particular

winniehell replied:
that change broke encoding empty tables which is not covered by upstream tests. I will fix that.

winniehell replied:
fixed upstream as well and also added tests there :heavy_check_mark:

winniehell commented on src/resources/pandoc/datadir/_json.lua:
```diff
@@ -26,7 +27,22 @@
 --
 
-local json = { _version = "0.1.2" }
+local json = { _version = "0.1.2-quarto" }
```
we can't keep the version number but I'm not sure if this is a good way to deal with it

cscheid replied:
This is fine.

winniehell commented on src/resources/pandoc/datadir/_json.lua:
```diff
@@ -72,7 +88,6 @@ local function encode_table(val, stack)
 
   stack[val] = true
 
-  local n = 0
```
unused

winniehell commented on src/resources/pandoc/datadir/_json.lua:
```diff
@@ -72,7 +88,10 @@ local function encode_table(val, stack)
 
-  local n = 0
+  if next(val) == nil then
+    return '[]'
```
this is consistent with previous behavior but an empty table could also be encoded as `{}`

cscheid replied:
The correspondence between JSON's `{}`,`[]`, `null` and Lua's `{}` and `nil` can never be done perfectly, so we should just maintain the previous behavior.

cscheid reviewed (commented):
This looks great, thank you.

We're still missing the changelog update, right?

winniehell committed:
Add changelog entries

winniehell commented on news/changelog-1.6.md:
```diff
@@ -29,6 +30,7 @@ All changes included in 1.6:
 - Fix `kbd` element styling on dark themes.
 - ([#10761](https://github.com/quarto-dev/quarto-cli/issues/10761)): Add support for `licence: CC0` to automatically link to Creative Commons licence [CC0 1.0](https://creativecommons.org/publicdomain/zero/1.0/).
 - ([#10817](https://github.com/quarto-dev/quarto-cli/issues/10817)): Ensure that user provided SCSS has precedent over quarto generated scss also for dark theme.
+- ([#11124](https://github.com/quarto-dev/quarto-cli/pull/11124)): Use stable order of GLightbox options
```
I have added two entries because I think there are two aspects of this change: one refactoring which may have unsurfaced side effects and one bugfix / enhancement. please feel free to delete either of them.

cscheid replied:
Ok. I'll remove the `html` format change.

winniehell replied:
most users will probably not notice this anyway :sweat_smile:

cscheid commented:
Merged - thanks for the PR and followup!
